ext/formatting.go

```diff
@@ -0,0 +1,878 @@
+	}
+}
+
+// stringFormatValidator implements the cel.ASTValidator interface allowing for static validation
+// of string.format calls.
+type stringFormatValidator struct{}
+
+// Name returns the name of the validator.
+func (stringFormatValidator) Name() string {
+	return "cel.lib.ext.validate.functions.string.format"
+}
+
+// Configure implements the ASTValidatorConfigurer interface and augments the list of functions to skip
+// during homogeneous aggregate literal type-checks.
+func (stringFormatValidator) Configure(config cel.MutableValidatorConfig) error {
+	functions := config.GetOrDefault(cel.HomogeneousAggregateLiteralExemptFunctions, []string{}).([]string)
+	functions = append(functions, "format")
+	return config.Set(cel.HomogeneousAggregateLiteralExemptFunctions, functions)
+}
+
+// Validate parses all literal format strings and type checks the format clause against the argument
+// at the corresponding ordinal within the list literal argument to the function, if one is specified.
+func (stringFormatValidator) Validate(env *cel.Env, _ cel.ValidatorConfig, a *ast.CheckedAST, iss *cel.Issues) {
+	root := ast.NavigateCheckedAST(a)
+	formatCallExprs := ast.MatchDescendants(root, matchConstantFormatStringWithListLiteralArgs)
+	for _, e := range formatCallExprs {
+		call := e.AsCall()
+		formatStr := call.Target().AsLiteral().Value().(string)
+		args := call.Args()[0].AsList().Elements()
+		formatCheck := &stringFormatChecker{
+			args:    args,
+			typeMap: a.TypeMap,
+		}
+		// use a placeholder locale, since locale doesn't affect syntax
+		_, err := parseFormatString(formatStr, formatCheck, formatCheck, "en_US")
+		if err != nil {
+			iss.ReportErrorAtID(getErrorExprID(e.ID(), err), err.Error())
+			continue
+		}
+		seenArgs := formatCheck.argsRequested
+		if len(args) > seenArgs {
+			iss.ReportErrorAtID(e.ID(),
+				"too many arguments supplied to string.format (expected %d, got %d)", seenArgs, len(args))
+		}
+	}
+}
+
+// getErrorExprID determines which list literal argument triggered a type-disagreement for the
+// purposes of more accurate error message reports.
+func getErrorExprID(id int64, err error) int64 {
+	fmtErr, ok := err.(formatError)
+	if ok {
+		return fmtErr.id
+	}
+	wrapped := errors.Unwrap(err)
+	if wrapped != nil {
+		return getErrorExprID(id, wrapped)
+	}
+	return id
+}
+
+// matchConstantFormatStringWithListLiteralArgs matches all valid expression nodes for string
+// format checking.
+func matchConstantFormatStringWithListLiteralArgs(e ast.NavigableExpr) bool {
+	if e.Kind() != ast.CallKind {
+		return false
+	}
+	call := e.AsCall()
+	// TODO: expose the OverloadID() method on the NavigableCallExpr to refine this check.
+	if !call.IsMemberFunction() || call.FunctionName() != "format" {
+		return false
+	}
+	formatString := call.Target()
+	if formatString.Kind() != ast.LiteralKind && formatString.Type() != cel.StringType {
+		return false
+	}
+	args := call.Args()
+	if len(args) != 1 {
```

Comment:
Out of curiosity, did you think there was anything wrong with checking for the arg length of the original call signature? It makes sense to eventually move back to the `overloadID` check instead of the function name (to avoid accidentally looking at other member functions that are also named `format`) but this seems fine to me.

Reply:
In the AST, the `format` is a receiver-style call: `<string>.format(<list>)`, so the call will have a `target` which will be the format string, and a single argument. During program planning, the call structure is normalized and the target is moved to the first argument. Hence, the subtle change here. Make sense?

Reply:
Right, I think I somehow misunderstood and thought you were wanting to change/replace the `len(args) != 1` check, sorry - this sounds good to me.